src/language_features/semantic_highlighting.rs

```diff
@@ -75,7 +69,32 @@ pub fn debug_scopes(meta: EditorMeta, ctx: &mut Context) {
     });
 }
 
-// TODO: Have an actual usable lookup table for faces
-fn get_face_for_scope(scope: &Vec<String>) -> String {
-    scope.iter().flat_map(|s| s.split(".")).join("_")
+pub fn make_scope_map(ctx: &mut Context) -> std::vec::Vec<std::string::String> {
+    let faces: HashMap<String, &String> = ctx.config.semantic_scopes.iter().map(|(k,v)| (k.replace("_", "."), v)).collect();
+
+    let scopes = ctx
+        .capabilities
+        .as_ref()
+        .and_then(|x| x.semantic_highlighting.as_ref())
+        .and_then(|x| x.scopes.as_ref());
+
+    if scopes == None {
+      return Vec::new();
+    }
+    let scopes = scopes.unwrap();
+
+    let faces: Vec<String> = scopes.iter().map(|scopes| {
+      for scope in scopes {
```

Comment:
nit: I think more idiomatic Rust could look more like

```rust
pub fn make_scope_map(ctx: &mut Context) -> std::vec::Vec<std::string::String> {
    let faces: HashMap<String, &String> = ctx
        .config
        .semantic_scopes
        .iter()
        .map(|(k, v)| (k.replace("_", "."), v))
        .collect();

    let scopes = ctx
        .capabilities
        .as_ref()
        .and_then(|x| x.semantic_highlighting.as_ref())
        .and_then(|x| x.scopes.as_ref());

    if scopes == None {
        return Vec::new();
    }
    let scopes = scopes.unwrap();

    map_scopes_to_faces(scopes, faces)
}

fn map_scopes_to_faces(
    scopes: &Vec<Vec<String>>,
    faces: HashMap<String, &String>,
) -> std::vec::Vec<std::string::String> {
    let find_face = |scope: &String| {
        scope
            .chars()
            .enumerate()
            .filter_map(|(i, x)| {
                if x == '.' {
                    // Trading extra HashMap lookups for creating intermediate collection seems to be fair.
                    faces.get(&scope[0..i])
                } else {
                    None
                }
            })
            .last()
            .map(|face| face.to_string())
    };
    scopes
        .iter()
        .map(|scopes| {
            scopes
                .iter()
                .find_map(find_face)
                .unwrap_or_else(|| String::new())
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn map_scopes_to_faces_should_map_unknown_scope_to_empty_face() {
        let scopes = vec![
            vec![String::from("just.noise")],
            vec![String::from("some.scope")],
        ];
        let mut faces: HashMap<String, &String> = Default::default();
        let some_face = String::from("some.face");
        faces.insert(String::from("some.another.scope"), &some_face);
        let faces = map_scopes_to_faces(&scopes, faces);
        assert_eq!(Some(&String::from("")), faces.get(1));
    }

    #[test]
    fn map_scopes_to_faces_should_map_scope_by_prefix() {
        let scopes = vec![
            vec![String::from("just.noise")],
            vec![
                String::from("some.non-matching.scope"),
                String::from("some.nested.scope"),
            ],
        ];
        let mut faces: HashMap<String, &String> = Default::default();
        let some_face = String::from("some.face");
        faces.insert(String::from("some.nested"), &some_face);
        let faces = map_scopes_to_faces(&scopes, faces);
        assert_eq!(Some(&String::from("some.face")), faces.get(1));
    }

    #[test]
    fn map_scopes_to_faces_should_map_scope_by_longest_prefix() {
        let scopes = vec![
            vec![String::from("just.noise")],
            vec![
                String::from("some.scope.matching.short.prefix"),
                String::from("some.nested.scope"),
            ],
        ];
        let mut faces: HashMap<String, &String> = Default::default();
        let some = String::from("some");
        let some_face = String::from("some.face");
        faces.insert(String::from("some"), &some);
        faces.insert(String::from("some.nested"), &some_face);
        let faces = map_scopes_to_faces(&scopes, faces);
        assert_eq!(Some(&String::from("some.face")), faces.get(1));
    }
}
```

I extracted `map_scopes_to_faces` for easier testing. Note the last test would fail with this code (as well as with original code). I don't know if this case is realistic enough or not for us to care about.

This is one of the ways to account for this edge case:

```rust
fn map_scopes_to_faces(
    scopes: &Vec<Vec<String>>,
    faces: HashMap<String, &String>,
) -> std::vec::Vec<std::string::String> {
    let find_face = |scope: &String| {
        scope
            .chars()
            .enumerate()
            .filter_map(|(i, x)| if x == '.' { Some(&scope[0..i]) } else { None })
            .enumerate()
            .filter_map(|(n, element)| faces.get(element).map(|face| (n, face)))
            .last()
            .map(|(n, face)| (n, face.to_string()))
    };
    scopes
        .iter()
        .map(|scopes| {
            scopes
                .iter()
                .filter_map(find_face)
                .max_by_key(|(n, _)| *n)
                .map(|(_, x)| x)
                .unwrap_or_else(|| String::new())
        })
        .collect()
}
```

Reply:
I'll use this code instead
Are you sure its supposed to behave like that in the last test case?

Reply:
I do see the logic in the longest name being the most specific. But i assumed they were ordered by the order they should be applied in.

I this makes sense when only mapping some scopes to faces though